<a href="https://colab.research.google.com/github/adchaturvedi/DataAnalytics/blob/main/DA_Exp3_LinearRegression_ClimateChange.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Name: Aditi Chaturvedi

UID:2019120011


Aim:Building Linear Regression model for given dataset- climate_change.csv

In [19]:
#Importing the librariesimport pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

In [20]:
#Reading the dataset
dataset = pd.read_csv("climate_change.csv")

In [21]:
dataset.head()

Year  Month    MEI     CO2      CH4      N2O   CFC-11   CFC-12        TSI  \
0  1983      5  2.556  345.96  1638.59  303.677  191.324  350.113  1366.1024   
1  1983      6  2.167  345.52  1633.71  303.746  192.057  351.848  1366.1208   
2  1983      7  1.741  344.15  1633.22  303.795  192.818  353.725  1366.2850   
3  1983      8  1.130  342.25  1631.35  303.839  193.602  355.633  1366.4202   
4  1983      9  0.428  340.17  1648.40  303.901  194.392  357.465  1366.2335   

   Aerosols   Temp  
0    0.0863  0.109  
1    0.0794  0.118  
2    0.0731  0.137  
3    0.0673  0.176  
4    0.0619  0.149

In [22]:
dataset.columns

Index(['Year', 'Month', 'MEI', 'CO2', 'CH4', 'N2O', 'CFC-11', 'CFC-12', 'TSI',
       'Aerosols', 'Temp'],
      dtype='object')

In [23]:
Q1 = dataset.quantile(0.25)
Q3 = dataset.quantile(0.75)
IQR = Q3 - Q1
#print(IQR)
dataset = dataset[~((dataset < (Q1 - 1.5 * IQR)) |(dataset > (Q3 + 1.5 * IQR))).any(axis=1)]

In [24]:
import statsmodels.api as sm
x = dataset[['MEI', 'CO2', 'CH4', 'N2O', 'CFC-11', 'CFC-12', 'TSI','Aerosols']]
y = dataset[['Temp']]
x2 = sm.add_constant(x)
est = sm.OLS(y,x2)
est2 = est.fit()
print(est2.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.703
Model:                            OLS   Adj. R-squared:                  0.692
Method:                 Least Squares   F-statistic:                     69.11
Date:                Tue, 22 Nov 2022   Prob (F-statistic):           2.36e-57
Time:                        09:27:09   Log-Likelihood:                 251.36
No. Observations:                 243   AIC:                            -484.7
Df Residuals:                     234   BIC:                            -453.3
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -60.8378     23.736     -2.563      0.0

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


Spliting the data into a training set, consisting of all the observations up to and including 2006, and a testing set consisting of the remaining years

In [25]:
df_train = dataset[dataset.iloc[:,0]<=2006]  
df_train.head()

Year  Month    MEI     CO2      CH4      N2O   CFC-11   CFC-12        TSI  \
29  1985     10 -0.140  343.08  1681.56  305.395  215.327  390.676  1365.5269   
30  1985     11 -0.050  344.40  1680.68  305.530  216.282  392.714  1365.6289   
31  1985     12 -0.293  345.82  1677.99  305.653  217.326  394.539  1365.6794   
32  1986      1 -0.307  346.54  1675.82  305.775  218.382  396.082  1365.6746   
33  1986      2 -0.191  347.13  1666.83  305.911  219.379  397.345  1365.5475   

    Aerosols   Temp  
29    0.0101 -0.008  
30    0.0097 -0.093  
31    0.0122 -0.002  
32    0.0146  0.121  
33    0.0158  0.065

In [26]:
df_test = dataset[dataset.iloc[:,0]>2006]  
df_test.head()

Year  Month    MEI     CO2      CH4      N2O   CFC-11   CFC-12  \
284  2007      1  0.974  382.93  1799.66  320.561  248.372  539.206   
285  2007      2  0.510  383.81  1803.08  320.571  248.264  538.973   
286  2007      3  0.074  384.56  1803.10  320.548  247.997  538.811   
287  2007      4 -0.049  386.40  1802.11  320.518  247.574  538.586   
288  2007      5  0.183  386.58  1795.65  320.445  247.224  538.130   

           TSI  Aerosols   Temp  
284  1365.7173    0.0054  0.601  
285  1365.7145    0.0051  0.498  
286  1365.7544    0.0045  0.435  
287  1365.7228    0.0045  0.466  
288  1365.6932    0.0041  0.372

linear regression model of traing set

In [27]:
#Setting the value for X and Y
x_train = df_train[['MEI', 'CO2', 'CH4', 'N2O', 'CFC-11', 'CFC-12', 'TSI','Aerosols']]
y_train = df_train['Temp']
x2_train = sm.add_constant(x_train)
est_train = sm.OLS(y_train,x2_train)
est2_train = est_train.fit()
print(est2_train.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.722
Model:                            OLS   Adj. R-squared:                  0.711
Method:                 Least Squares   F-statistic:                     68.15
Date:                Tue, 22 Nov 2022   Prob (F-statistic):           3.37e-54
Time:                        09:27:09   Log-Likelihood:                 229.49
No. Observations:                 219   AIC:                            -441.0
Df Residuals:                     210   BIC:                            -410.5
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -51.0320     24.469     -2.086      0.0

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


**Problem 1.1**
The value of R-squared : 0.722


**Problem 1.2**
Which variables are significant in the model?

Ans. : MEI, CO2, TSI ,Aerosols (p-value <0.05)

In [28]:
#Setting the value for X and Y
x_test = df_test[['MEI', 'CO2', 'CH4', 'N2O', 'CFC-11', 'CFC-12', 'TSI',
       'Aerosols']]
y_test = df_test['Temp']

In [29]:
x_train.size


1752

In [30]:
y_train.size


219

In [31]:
x_test.size


192

In [32]:
y_test.size

24

In [33]:
from sklearn.linear_model import LinearRegression

In [34]:
mlr = LinearRegression()
mlr.fit(x_train,y_train)

LinearRegression()

In [35]:
print("Intercept: ", mlr.intercept_)
print("Coefficients:")
list(zip(x_train, mlr.coef_))

Intercept:  -51.031969159858036
Coefficients:


[('MEI', 0.06223569777302381),
 ('CO2', 0.0049606987940408465),
 ('CH4', -0.00038810727802363575),
 ('N2O', 0.0018262419311547184),
 ('CFC-11', -0.0011344993284555694),
 ('CFC-12', 0.0014013277032073551),
 ('TSI', 0.03604734063953119),
 ('Aerosols', -8.435947559286046)]

In [36]:
from scipy.stats import pearsonr
list1 = df_train['MEI']
list2 = df_train['N2O']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: -0.062


In [37]:
from scipy.stats import pearsonr
list1 = df_train['CO2']
list2 = df_train['N2O']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.975


In [38]:
from scipy.stats import pearsonr
list1 = df_train['CH4']
list2 = df_train['N2O']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.890


In [39]:
from scipy.stats import pearsonr
list1 = df_train['CFC-11']
list2 = df_train['N2O']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.327


In [40]:
from scipy.stats import pearsonr
list1 = df_train['CFC-12']
list2 = df_train['N2O']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.865


In [41]:
from scipy.stats import pearsonr
list1 = df_train['TSI']
list2 = df_train['N2O']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.160


In [42]:
from scipy.stats import pearsonr
list1 = df_train['Aerosols']
list2 = df_train['N2O']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: -0.661


**Problem 2.1**
Which of the following is the simplest correct explanation for this contradiction?

I. Climate scientists are wrong that N2O and CFC-11 are greenhouse gases - this regression analysis constitutes part of a disproof.

II. There is not enough data, so the regression coefficients being estimated are not accurate.

III. All of the gas concentration variables reflect human development - N2O and CFC.11 are correlated with other variables in the data set.

Ans:All of the gas concentration variables reflect human development - N2O and CFC.11 are correlated with other variables in the data set.


##Correlation
Which of the following independent variables is N2O highly correlated with (absolute correlation greater than 0.7)? 

ans. b) CO2 c) CH4  f) CFC.12

In [43]:
from scipy.stats import pearsonr
list1 = df_train['MEI']
list2 = df_train['CFC-11']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: -0.166


In [44]:
from scipy.stats import pearsonr
list1 = df_train['CO2']
list2 = df_train['CFC-11']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.342


In [45]:
from scipy.stats import pearsonr
list1 = df_train['CH4']
list2 = df_train['CFC-11']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.628


In [46]:
from scipy.stats import pearsonr
list1 = df_train['N2O']
list2 = df_train['CFC-11']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.327


In [47]:
from scipy.stats import pearsonr
list1 = df_train['CFC-12']
list2 = df_train['CFC-11']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.749


In [48]:
from scipy.stats import pearsonr
list1 = df_train['TSI']
list2 = df_train['CFC-11']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: 0.257


In [49]:
from scipy.stats import pearsonr
list1 = df_train['Aerosols']
list2 = df_train['CFC-11']
 
# Apply the pearsonr()
corr, _ = pearsonr(list1, list2)
print('Pearsons correlation: %.3f' % corr)

Pearsons correlation: -0.231


Which of the following independent variables is CFC.11 highly correlated with? Select all that apply.

ans: CFC.12 

In [50]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(x_train, y_train)

LinearRegression()

In [51]:
y_pred = regressor.predict(x_test)

In [52]:
df_train.drop(['CH4', 'N2O', 'CFC-11', 'CFC-12'],axis=1)

Year  Month    MEI     CO2        TSI  Aerosols   Temp
29   1985     10 -0.140  343.08  1365.5269    0.0101 -0.008
30   1985     11 -0.050  344.40  1365.6289    0.0097 -0.093
31   1985     12 -0.293  345.82  1365.6794    0.0122 -0.002
32   1986      1 -0.307  346.54  1365.6746    0.0146  0.121
33   1986      2 -0.191  347.13  1365.5475    0.0158  0.065
..    ...    ...    ...     ...        ...       ...    ...
279  2006      8  0.759  380.45  1365.7067    0.0041  0.482
280  2006      9  0.793  378.92  1365.8419    0.0043  0.425
281  2006     10  0.892  379.16  1365.8270    0.0044  0.472
282  2006     11  1.292  380.18  1365.7039    0.0049  0.440
283  2006     12  0.951  381.79  1365.7087    0.0054  0.518

[219 rows x 7 columns]

In [53]:
x_train = df_train[['MEI','N2O', 'TSI','Aerosols']]
y_train = df_train['Temp']
x2_train = sm.add_constant(x_train)
est_train = sm.OLS(y_train,x2_train)
est2_train = est_train.fit()
print(est2_train.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.706
Model:                            OLS   Adj. R-squared:                  0.701
Method:                 Least Squares   F-statistic:                     128.7
Date:                Tue, 22 Nov 2022   Prob (F-statistic):           8.92e-56
Time:                        09:27:09   Log-Likelihood:                 223.51
No. Observations:                 219   AIC:                            -437.0
Df Residuals:                     214   BIC:                            -420.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -53.1366     23.146     -2.296      0.0

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


 ##3.1

 
 Enter the coefficient of N2O in this reduced model:  0.0217


 R-squared: 0.706

In [54]:
x_train = df_train[['MEI', 'CO2', 'TSI','Aerosols']]
y_train = df_train['Temp']
x2_train = sm.add_constant(x_train)
est_train = sm.OLS(y_train,x2_train)
est2_train = est_train.fit()
print(est2_train.summary())

                            OLS Regression Results                            
Dep. Variable:                   Temp   R-squared:                       0.712
Model:                            OLS   Adj. R-squared:                  0.707
Method:                 Least Squares   F-statistic:                     132.2
Date:                Tue, 22 Nov 2022   Prob (F-statistic):           1.13e-56
Time:                        09:27:09   Log-Likelihood:                 225.63
No. Observations:                 219   AIC:                            -441.3
Df Residuals:                     214   BIC:                            -424.3
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        -57.0416     22.980     -2.482      0.0

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:142: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


Conclusion:
From the analysis of the results obtained from tests performed , I came to know that the variables 'CFC_12', 'Aerosols', 'CFC_11', 'MEI', 'TSI' are significant in the model i.e. their p-value is below 0.05.
All of the gas concentration variables reflect human development - N2O and CFC.11 are correlated with other variables in the data set. To verify this, I tried removing/adding some of the variables and observed different values of coefficient indicating its dependence on values of coefficient of other variables.